### PR TITLE
Categorization

### PR DESCRIPTION
**Pull request by sejkimm**
### **1. Add Route /category** 
**1.1 Add /html/CategoryList.html**
-  It renders /category

**1.2 Add func _ListCategory() in (app.go)**
- It controls /html/CategoryList.html

**1.3 Add func _getCategoryList() in (dataprocess.go)**
- Get category list form DB and returns to _ListCategory()

### **2. add Route /category/{cat}** 
**2.1 Add /html/Category.html**
-  It renders /category/{cat}. It is similar with /html/index.html but it does not cotain paging and submit code href.

**2.2 Add func _getCategoryCode() in (app.go)**
- It controls /html/Category.html

**2.3 Add func _getPostListFromCategory() in (dataprocess.go)**
- Get posts from DB and returns to _getCategoryCode(). It is similar with _getPostList() in dataprocess.go but it returns specific code from DB that matches category.

### **3. modify DB Scheme**
Check Line19~32 in dataprocess.go

### **Things to check meticulously**
- function _ListCategory in (app.go), Line 163~167. 
- /html/categorylist.html, Line 25~27.

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -151,3 +154,49 @@ func _registerComment(w http.ResponseWriter, r *http.Request) {
 	}
 
 }
+
+// Listing All Categories
+func _ListCategory(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+
+	if err == nil {
```
Comment (joona0825) [inline app.go:162]: Variable err is not declared here

Comment (hellojunha) [inline app.go:165]: Variable err is not declared

File: app.go [hunk 2]
```diff
@@ -151,3 +154,49 @@ func _registerComment(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+}
+
+// Render codes matches category in DB
+func _getCategoryCode(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	page := vars["cat"]
+
+	list = getPostListFromCategory(cat)
```
Comment (hellojunha) [inline app.go]: No pagination needed?

File: app.go [hunk 3]
```diff
@@ -151,3 +154,49 @@ func _registerComment(w http.ResponseWriter, r *http.Request) {
...
 
 }
+
+// Listing All Categories
+func _ListCategory(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+
+	if err == nil {
+		categroies := _getCategoryList() // returns string type slice 
+		t, err := template.ParseFiles(wd + "/html/categoryList.html")
```
Comment (hellojunha) [inline app.go]: Method for getting template file has changed. Check out the latest commit.

File: app.go [hunk 4]
```diff
@@ -151,3 +154,49 @@ func _registerComment(w http.ResponseWriter, r *http.Request) {
...
+}
+
+// Render codes matches category in DB
+func _getCategoryCode(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	page := vars["cat"]
+
+	list = getPostListFromCategory(cat)
+
+	t, err := template.ParseFiles(wd + "/html/index.html")
```
Comment (hellojunha) [inline app.go]: Method for getting template file has changed. Check out the latest commit.

File: html/post.html [hunk 5]
```diff
@@ -11,7 +11,15 @@
     };
 </script>
 
+
 <form id="code_form" action="/post/register" method="post">
+    <!-- Dropdown to select Category-->
+    <select name="category">
+        <option value="LinkedList">LinkedList</option>
```
Comment (hellojunha) [inline html/post.html]: hard code?

File: dataprocess.go [hunk 6]
```diff
@@ -20,9 +20,17 @@ type Post struct {
 	Id int
 	Dt time.Time
 	Code string
+	Category []Category // Add relationship in DB
```
Comment (hellojunha) [inline dataprocess.go]: Why array?

File: dataprocess.go [hunk 7]
```diff
@@ -94,6 +102,67 @@ func getPostList(page int) ([]Post, int) {
...
+	defer db.Close()
+
+	cats := make([]Category, 0)
+
+	rows, err := db.Query("select category from study_category") // Check this Query
+
+	if err == nil {
+		defer rows.Close()
+		for rows.Next() {
+			err := rows.Scan(&category)
```
Comment (hellojunha) [inline dataprocess.go]: Variable category is not declared

File: dataprocess.go [hunk 8]
```diff
@@ -94,6 +102,67 @@ func getPostList(page int) ([]Post, int) {
...
+		log.Println(err.Error())
+	}
+
+	// log.Printf("getting categories")
+
+	return cats
+}
+
+// similar as func getPostList() but Query is little different, and not paging  
+func getPostListFromCategory(string cat) [] {
```
Comment (hellojunha) [inline dataprocess.go]: return type?

**Review by hellojunha — CHANGES_REQUESTED** [review]

**Comment by hellojunha** [comment]
Please upload sql query I need to execute (study_category table). And you don't need too much comments in code as you wrote. I personally do not prefer writing comment because it means the code is hard to read and understand.

File: app.go (hunk 1 above)
Comment (sejkimm) [reply app.go:165]: Fixed in App.go Line 161   d3e72a5

File: app.go (hunk 2 above)
Comment (sejkimm) [reply app.go]: All categories have fewer than 10 posts.

File: app.go (hunk 3 above)
Comment (sejkimm) [reply app.go]: Fixed

File: app.go (hunk 4 above)
Comment (sejkimm) [reply app.go]: Fixed

File: html/post.html (hunk 5 above)
Comment (sejkimm) [reply html/post.html]: Fixed

File: dataprocess.go (hunk 6 above)
Comment (sejkimm) [reply dataprocess.go]: Fixed. Now string

File: dataprocess.go (hunk 7 above)
Comment (sejkimm) [reply dataprocess.go]: Fixed

File: dataprocess.go (hunk 8 above)
Comment (sejkimm) [reply dataprocess.go]: Fixed

**Comment by sejkimm** [comment]
Thanks for your review. 
I fixed several errors and It now works properly on my local environment.

SQL query for my commit is below 
`CREATE TABLE study_category ( category_id INT(11) NOT NULL AUTO_INCREMENT, category TINYTEXT NOT NULL, Week INT(11) NOT NULL, PRIMARY KEY(category_id));`

and might need one more query 
`ALTER TABLE study_post ADD COLUMN category TINYTEXT NOT NULL`

Also I removed every comment I wrote.

**Pull request merged**